pkg/operation/botanist/clusteridentity/clusteridentity.go

```diff
@@ -0,0 +1,97 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package clusteridentity
+
+import (
+	"context"
+	"fmt"
+
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	v1beta1constants "github.com/gardener/gardener/pkg/apis/core/v1beta1/constants"
+	"github.com/gardener/gardener/pkg/operation/botanist/component"
+	kutil "github.com/gardener/gardener/pkg/utils/kubernetes"
+
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type shootClusterIdentity struct {
+	clusterIdentity *string
+	seedNamespace   string
+	namespace       string
+	name            string
+	uid             string
+
+	gardenClient client.Client
+
+	logger *logrus.Entry
+}
+
+// New creates new instance of Deployer for Shoot cluster identity
+func New(
+	clusterIdentity *string,
+	name, namespace, seedNamespace, uid string,
+	gardenClient client.Client,
+	logger *logrus.Entry,
+) component.Deployer {
+	return &shootClusterIdentity{
+		clusterIdentity: clusterIdentity,
+		name:            name,
+		namespace:       namespace,
+		uid:             uid,
+		seedNamespace:   seedNamespace,
+		gardenClient:    gardenClient,
+		logger:          logger,
+	}
+}
+
+func (c *shootClusterIdentity) Deploy(ctx context.Context) error {
+	if c.clusterIdentity == nil {
+		clusterIdentity, err := c.computeShootClusterIdentity(ctx)
+		if err != nil {
+			return err
+		}
+
+		shoot := &gardencorev1beta1.Shoot{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      c.name,
+				Namespace: c.namespace,
+			},
+		}
+
+		patch := []byte(fmt.Sprintf(`{"status": {"clusterIdentity": "%s"}}`, clusterIdentity))
+		return c.gardenClient.Status().Patch(ctx, shoot, client.RawPatch(types.StrategicMergePatchType, patch))
+	}
+	return nil
+}
+
+// Destroy returns nil
+func (c *shootClusterIdentity) Destroy(ctx context.Context) error {
+	return nil
+}
+
+// computeShootClusterIdentity computes the identity of a Shoot cluster and returns it.
+func (c *shootClusterIdentity) computeShootClusterIdentity(ctx context.Context) (string, error) {
+	gardenClusterIdentity := &corev1.ConfigMap{}
+	if err := c.gardenClient.Get(ctx, kutil.Key(metav1.NamespaceSystem, v1beta1constants.ClusterIdentity), gardenClusterIdentity); err != nil {
```

Comment:
Is the garden identity read for every shoot identity creation? It is immutable and should be constant for the whole lifetime of the gardenlet.

Reply:
It is validated, that the identity is not changed in shoot and seed objects, but what about the configmap(s) in the kube-system namespace. Here it should also not be possible to change the `cluster-identity` field.

May be I missed it, but is the shoot identity part of the cluster info object in the shoot namespace of the seed?

Reply:
> Is the garden identity read for every shoot identity creation? It is immutable and should be constant for the whole lifetime of the gardenlet.

Agreed, it could be read once when the gardenlet starts.

> May be I missed it, but is the shoot identity part of the cluster info object in the shoot namespace of the seed?

I don't think so, no. The identities are only stored in the kube-system namespaces. Probably it should be also stored in a `.status.identity` field in the shoot status.

> It is validated, that the identity is not changed in shoot and seed objects, but what about the configmap(s) in the kube-system namespace. Here it should also not be possible to change the cluster-identity field.

Hm, good point. How to validate it? The only way is to have webhooks, but this is also again a lot of effort. I suggest to put the shoot identity in the new `.status.identity` (immutable) field and use it as the source of truth.

Reply:
Hello @mandelsoft 

> Is the garden identity read for every shoot identity creation? It is immutable and should be constant for the whole lifetime of the gardenlet.

Yes I agree. 

> May be I missed it, but is the shoot identity part of the cluster info object in the shoot namespace of the seed?

It is part of the `Shoot.Status.clusterIdentity` and it is in the `Cluster` resource in the `Seed`, although for the `Cluster` nothing extra was implemented, this is done by synching it in the end of the `Shoot` reconciliation. Wouldn't this be enough?

>> It is validated, that the identity is not changed in shoot and seed objects, but what about the configmap(s) in the kube-system namespace. Here it should also not be possible to change the cluster-identity field.

> Hm, good point. How to validate it? The only way is to have webhooks, but this is also again a lot of effort. I suggest to put the shoot identity in the new .status.identity (immutable) field and use it as the source of truth.

The way it works now for the Seed is, the main source of truth is the `status.clusterIdentity`, but if it is missing we lookup for a `ConfigMap`, if  a ConfigMap doesn't exist as well, we compute the clusterIdentity. This is done only for the Seed. For the Shoot we are using only the clusterIdentity field in the status and if it is missing we compute it.

Reply:
Sounds great, thanks Svilen.